File: Heapsort/heapSort.py
def get_input():
    """
    get input from user
    """
    input_str = input("Enter elements to be sorted: ")
    try:
        elements = [int(e) for e in input_str.split()] #make a list of integers from input string
    except ValueError:
        print("Please enter a list of integers only, seperated by a space!!")
    return elements

def max_heapify(thelist, lst_size: int, idx: int):
    """
    Building maximum heap at index passed as argument
    """
    largest = idx
    left_child = 2 * idx + 1
    right_child = 2 * idx + 2

    #comparing left child and largest
    if left_child < lst_size and thelist[left_child] > thelist[largest]:
        largest = left_child

    #comparing right child and largest
    if right_child < lst_size and thelist[right_child] > thelist[largest]:
        largest = right_child

    if largest != idx: #if largest is changed
        thelist[idx], thelist[largest] = thelist[largest], thelist[idx] #swapping
        max_heapify(thelist, lst_size, largest)

def build_max_heap(thelist, lst_size: int):
    """
    Building maximum heap from the list passed as argument
    """
    for curr_idx in range(lst_size // 2 - 1, -1, -1): #loop from higher index to lower index
        max_heapify(thelist, lst_size, curr_idx)

def heap_sort(thelist):
    """
    Heap sort Algorithm
    """
    if len(thelist) == 0:
        print("Empty list!!")

    elif len(thelist) == 1:
        print("Only one element!!")

    else:
        build_max_heap(thelist, len(thelist))

        for curr_idx in range(len(thelist) -1, 0, -1): #loop from higher index to lower index
            thelist[curr_idx], thelist[0] = thelist[0], thelist[curr_idx] #swapping
            max_heapify(thelist, curr_idx, 0)

if __name__ == '__main__':
    input_list = get_input()
    heap_sort(input_list)
    print(*input_list, sep = ", ")

File: Heapsort/C++/heapsort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

void max_heapify(std::vector<int>& arr, int i, int size_)
{
    int largest, l = (2*i) + 1, r = l + 1;

    if(l < size_ && arr[l] > arr[i])
        largest = l;
    else
        largest = i;

    if(r < size_ && arr[r] > arr[largest])
        largest = r;

    if(largest != i)
    {
        std::swap(arr[i], arr[largest]);
        max_heapify(arr, largest, size_);
    }
}

void build_max_heap(std::vector<int>& arr)
{
    for(int i = (arr.size() / 2); i >= 0; i--)
    max_heapify(arr, i, arr.size());
}

void heap_sort(std::vector<int>& arr)
{
   build_max_heap(arr);
   int sz = arr.size();
   for(int i = arr.size() - 1; i > 0; i--)
   {
        std::swap(arr[0], arr[i]);
        sz--;
        max_heapify(arr, 0, sz);
    }
}

int main()
{
    std::vector<int> arr = {4, 1, 3, 2, 16, 9, 10, 14, 8, 7};
    heap_sort(arr);
    
    for(int i = 0; i < arr.size(); i++)
    {
         std::cout << arr[i] << " ";
    }
    std::cout << "\n";
    return 0;
}
